cmake/kokkos_tribits.cmake

```diff
@@ -280,6 +280,10 @@ FUNCTION(KOKKOS_LINK_TPLS LIBRARY_NAME)
     TARGET_INCLUDE_DIRECTORIES(${LIBRARY_NAME} PUBLIC ${HPX_INCLUDE_DIRS})
   ENDIF()
 
+  IF (KOKKOS_ENABLE_PTHREAD)
```

Comment:
How does this interact with tribits?

Reply:
@mhoemmen AFAIK, Pthread is only a TPL for Kokkos. There is no standard Trilinos Pthread TPL. Thus we can change the TPL behavior here and it won't break anything else in Trilinos.

The preferred way to do this is to use standard CMake `find_package(Threads)` and link to the target `Threads::Threads`.  We could change the Kokkos TPL file to do exactly this. Trilinos is based on CMake 3 - so even though TriBITS is not depending on transitive flags, the transitive flags will still happen.

The problem with depending on an imported target like `Threads::Threads` is that you have *re-import* the target in downstream packages. We can make this work for standalone Kokkos. @mhoemmen @bartlettroscoe Do you know if we can make an imported CMake target inside a TriBITS FindTPL file?

Reply:
Honestly... we might have a conversation on whether the time has come to make Kokkos a Trilinos TPL.

Reply:
The logic in Kokkos Trilinos right now would add `-lpthread`, not `-pthread`.

Reply:
@jjwilke asked:

>  @bartlettroscoe Do you know if we can make an imported CMake target inside a TriBITS FindTPL file?

you can write the `FindTPL<TPLName>.cmake` file any way you want.  See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-use-find-package-for-a-tribits-tpl

Reply:
@jjwilke said:

> Honestly... we might have a conversation on whether the time has come to make Kokkos a Trilinos TPL.

TPLs are for things that hardly ever change and there is not standard process for installing TPLs so this would be a **major** disruption to users of Trilinos (and Trilinos PR testing) and everything else.

What is so bad about maintaining two CMake build systems?  I thought that was the goal of that major Kokkos CMake refactoring that was just completed?

If there are some changes or other enhancements to TriBITS that would make this easier, then let's talk.

Reply:
@bartlettroscoe The build system for Kokkos is much easier to maintain if we have a FindKokkos.cmake inside Trilinos that calls `find_package(Kokkos)` and every package that needs Kokkos just calls `target_link_libraries(Kokkos::kokkos)`.  All of the linker and compiler flags now magically get set. And now we just maintain the build system for non-Trilinos users.  This also solves the problem that some have where they have Spack trees that depend on both Trilinos and Kokkos.  

I definitely understand this would be initially disruptive. I guess the question is whether this would be "long-term" disruptive or people would adjust quickly once they updated build scripts. I'm not in a good position to answer that. At the very least, making a Kokkos a TPL that doesn't need to use TriBITS would reduce complexity on our end.

Reply:
@bartlettroscoe Maybe we can delay this for now if this works alrready.  I couldn't parse my way through TriBITS to figure this out. If I create an imported target `Threads::Threads` and set `TPL_PTHREAD_LIBRARIES=Threads::Threads`... will `TRIBITS_ADD_LIBRARY(Kokkos)`  eventually call `TARGET_LINK_LIBRARIES(KOKKOS Thread::Threads)`?

We need the TPL dependence to add the `-pthread` *compile* option - and to do so "publicly".  It's not immediately obvious to me how we use the current Tribits TPL system to force compiler flags on downstream projects.

Are we good if we create imported targets like `Threads::Threads` and set `TPL_PTHREAD_LIBRARIES=Thread::Threads` and rely on CMake 3 target/property magic?

Reply:
> If I create an imported target Threads::Threads and set `TPL_PTHREAD_LIBRARIES=Threads::Threads`... will `TRIBITS_ADD_LIBRARY(Kokkos) `eventually call `TARGET_LINK_LIBRARIES(KOKKOS Thread::Threads)`?

Yes

> We need the TPL dependence to add the `-pthread` compile option - and to do so "publicly". 

That would need a TriBITS extension but that is consistent with `target_compile_definitions()` so there is no reason not to add that.

> Are we good if we create imported targets like `Threads::Threads` and set `TPL_PTHREAD_LIBRARIES=Thread::Threads` and rely on CMake 3 target/property magic?

It should.  If you think about it, that is how linking between internally build TriBITS packages work.